Validates updated parameters.

NigelOrr-Imenco opened the pull request:
### Closes/fixes/resolves issue(s)?
Action from meeting

### What was added/changed/fixed?
Adds workflow and action to validate updated parameters.
Adds workflow readme with basic usage information to allow script to be run offline to validate changes

### Related issue(s)? [Optional]
N/A

### Others [Optional]
N/A

### Checklist (after created PR)
- [ ] Check branch naming
- [ ] Remember to check universal design
- [ ] Added reviewers
- [ ] Added assignee(s)
- [ ] Added label(s)
- [ ] Reference to issue
- [ X ] I have performed a self-review of my own code.
- [ X ] My code is easy to read, and comments are added where needed.

### Documentation
- Need to update documentation? (Y/n) No - included

NigelOrr-Imenco committed:
Adds workflow and action to validate updated parameters.
Adds readme with basic usage information to allow script to be run offline to validate changes

njudell reviewed (dismissed):
Approved as demonstrated by Nigel

kndale commented:
@fschill @KimTommyUnplg @spacciniwsense @paul-randall-sonardyne 

Please review

paul-randall-sonardyne reviewed (dismissed):
Seems sensible - checking for duplicate ids and doing the protobuf generation.

NigelOrr-Imenco committed:
Removes surplus comma's from parameters.json

kndale reviewed (dismissed):
Looks good to me

njudell reviewed (commented):
Nigel:  I also did this in my PR 13.
Should be fine?

NigelOrr-Imenco commented:
As your PR hadn't been accepted yet, I couldn't pull it, so added in to my open one, I think it should merge OK. I needed to clear it out so I could update the validate script to auto-renumber and add default access permissions. There may have been a cleaner way to do that, happy to hear from the github professionals for next time!

NigelOrr-Imenco committed:
Adds --rewrite_default_access and --rewrite_auto_number_id options
As requested at previous meeting.
These options will overwrite the specified --file when changes are required
Updates readme.md accordingly

NigelOrr-Imenco commented:
Updated validation script, updated parameters.json using it to add default access for parameters where access was not specified

NigelOrr-Imenco committed:
Updates .gitignore to allow auto-generated parameters.csv
Adds initial csv and proto which should be overwritten automatically

NigelOrr-Imenco commented:
There may be a better solution to this, to allow parameters.csv to only be overwritten by the validation script, but have removed it from .gitignore so validation can complete without error.

fschill reviewed (dismissed):
Looks good; I ran it locally on Ubuntu 22.04, it ran after following instructions in Readme. I did not review in detail that parameters.json is still the same as the diff is big, but it looks like it's just formatting and adding the missing flags, which looks ok to me.

cbwsense reviewed (dismissed):
It looks good. I ran it locally with and without the --rewrite_default_access option, it works fine in both cases. When the --rewrite_default_access option is used, if the "access" field is missing, then it is added correctly with default values ​​and the parameters.json is overwritten. I noticed that the overwrite operation changes the indentation of the parameters.json and that the Celsius symbol doesn't seem to be encoded correctly for "id":25 description in both parameters.json and parameters.csv

NigelOrr-Imenco committed:
Merge branch 'main' into main

njudell commented:
Funny, I don't see an approved button.....

njudell reviewed (dismissed):
Approved as per my previous reviews.

NigelOrr-Imenco commented:
Closed as discussed

NigelOrr-Imenco committed:
Delete auto-generated csv file - housekeeping

NigelOrr-Imenco committed:
Delete auto-generated parameters.proto file - housekeeping

NigelOrr-Imenco commented:
Updated to remove auto-generated csv and proto file, which I think is what is needed to add the validation without trampling on other files.

njudell reviewed (approved):
Finally!

paul-randall-sonardyne reviewed (approved):
As per original review - all seems sensible.
A minor remark - the ID_UNASSIGNED value may need increasing in the future if the parameter count grows but it is fine for now.

njudell reviewed (commented):
The new workflow uses a form like:
python -m pip install.....

This does not work on some Linux distros, like Ubuntu 24.04. Any chance of adding error handling, or at least warnings, or alternative forms in comments?
(I've just been bitten by this problem trying to get pre-release reference code running on various platforms)

NigelOrr-Imenco commented:
> The new workflow uses a form like: python -m pip install.....
> 
> This does not work on some Linux distros, like Ubuntu 24.04. [...]

What is a suitable form? [The pip documentation](https://pip.pypa.io/en/stable/cli/pip_install/) recommends the above, are you sure it's failing and not the following 
`pip install -r .github/actions/validate_params/requirements.txt` 
line which I'd expect should be swapped to 
`python -m pip install -r .github/actions/validate_params/requirements.txt` 
to make it robust

The only potential problem I could see is if the Ubuntu 24.04 installation has python2 as the default, if you run `python --version` what is displayed?

njudell commented:
version is 3.12.3
Trying to install numpy that way yields:
python -m pip install numpy
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

    See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

Trying to do the install you listed:
 python -m pip install -r .github/actions/validare_params/requirements.txt
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

    See /usr/share/doc/python3.12/README.venv for more information.

I'm not sure how important this really is, in the big scheme of things.

NigelOrr-Imenco commented:
> Trying to do the install you listed: python -m pip install -r .github/actions/validare_params/requirements.txt error: externally-managed-environment
> 
> × This environment is externally managed ╰─> To install Python packages system-wide, try apt install python3-xyz, where xyz is the package you are trying to install.
 
That is trying to protect the user from updating system python libraries. To avoid that I run python with conda or venv so the python environment is sandboxed, which is [recommended](https://stackoverflow.com/questions/75608323/how-do-i-solve-error-externally-managed-environment-every-time-i-use-pip-3).

An alternative (e.g. if you are only running this python on the machine and don't need other python configurations) is to run pip with the option `--break-system-packages` or the other overrides on the above stackoverflow, which I do use if I want a system wide change, but wouldn't want to put in a script being run by others.

You are welcome to add informative text to readme's if you see a value in it, otherwise I think the error message is sufficient to point unfamiliar users to the explanation and solution.

njudell commented:
Agreed. I've spent way too long using 22.04

fschill commented:
re. installing python packages with pip, as Nigel already mentioned, the recommended way is to use Virtual Environments (venv), and install the packages only within the project, not system-wide.

For those not familiar, the commands are (in the project folder, for Linux):
```
python -m venv venv/              # set up the virtual environment
source venv/bin/activate         # activate the virtual environment
pip install -r requirements.txt # installs the packages in the venv only
```
when you're done, you can leave the virtual environments with ```deactivate```

If you later want to run it again, you can enter it again by only running the ```source venv/bin/activate``` command.

fschill commented:
could I suggest adding the following line to the readme, after "install python"?

* set up virtual environment in the project folder (```python3 -m venv venv/ && source venv/bin/activate``` )

cbwsense reviewed (approved):
As per previous review
